src/servers/src/http/jaeger.rs

```diff
@@ -0,0 +1,1244 @@
+            let mut process_id_to_process = HashMap::new();
+            for (service_name, process_id) in processes.into_iter() {
+                process_id_to_process.insert(
+                    process_id,
+                    Process {
+                        service_name,
+                        tags: vec![],
+                    },
+                );
+            }
+            trace.processes = Some(process_id_to_process);
+        }
+        traces.push(trace);
+    }
+
+    traces
+}
+
+// Construct Jaeger operations from records.
+fn operations_from_records(records: HttpRecordsOutput, contain_span_kind: bool) -> Vec<Operation> {
+    let mut operations = Vec::with_capacity(records.total_rows);
+    for row in records.rows.into_iter() {
+        let mut row_iter = row.into_iter();
+        if let Some(JsonValue::String(operation)) = row_iter.next() {
+            let mut operation = Operation {
+                name: operation,
+                span_kind: None,
+            };
+            if contain_span_kind {
+                if let Some(JsonValue::String(span_kind)) = row_iter.next() {
+                    operation.span_kind = Some(normalize_span_kind(span_kind.as_str()).to_string());
+                }
+            } else {
+                // skip span kind.
+                row_iter.next();
+            }
+            operations.push(operation);
+        }
+    }
+    operations
+}
+
+// FIXME(zyy17): It may encounter performance issue when the trace table is too large.
+fn get_services_sql(db: &str, trace_table_name: &str) -> String {
+    format!(
+        "SELECT DISTINCT(service_name) FROM {}.{}",
+        db, trace_table_name
+    )
+}
+
+// FIXME(zyy17): `trace_id` is high-cardinality column, it should use bloomfilter index.
+fn get_trace_sql(db: &str, trace_table_name: &str, trace_id: &str) -> String {
+    format!(
+        "SELECT trace_id, timestamp, duration_nano, service_name, span_name, span_id, span_attributes FROM {}.{} WHERE trace_id = '{}'",
+        db, trace_table_name, trace_id
+    )
+}
+
+// FIXME(zyy17): It may encounter performance issue when the trace table is too large.
+fn get_operations_sql(
+    db: &str,
+    trace_table_name: &str,
+    service_name: &str,
+    span_kind: Option<&str>,
+) -> String {
+    let mut query = format!(
+        "SELECT span_name, span_kind FROM {}.{} WHERE service_name = '{}'",
+        db, trace_table_name, service_name
+    );
+    if let Some(span_kind) = span_kind {
+        query = format!(
+            "{} AND span_kind = '{}'",
+            query,
+            normalize_span_kind(span_kind)
+        );
+    }
+    query
+}
+
+fn query_trace_sql(trace_table_name: &str, query_params: QueryTraceParams) -> String {
+    let mut query = format!(
+        "SELECT trace_id, timestamp, duration_nano, service_name, span_name, span_id, span_attributes FROM {}.{} WHERE service_name = '{}'",
+        query_params.db, trace_table_name, query_params.service_name
+    );
```

Comment:
sql injection warning. we should use parser to validate the query generated from this method contains only 1 statement.

Reply:
I use the DataFrame APIs to refactor all the query process. It may not have SQL injection problem in theory.